# Table of Contents
## 0) Importing Libraries & Datasets
## 1) Task 4.5 - Data Consistency Checks
## 2) Checking 'df_prods'
## 3) Check for Mixed-Type Data
## 4) Finding Missing Values
## 5) Address Missing Values
## 6) Check for Duplicate Values
## 7) Exporting

# 0) Importing Libraries & Datasets# Importing

In [1]:
# import numpy, pandas, os
import numpy as np
import pandas as pd
import os

In [2]:
# create file path 'path' for future access
path=r'/Users/sjpeterson76/Desktop/CareerFoundry/Data Immersion/Achievement 4/04-25-2021- Instacart Basket Analysis'

In [3]:
# import df_prods
df_prods=pd.read_csv(os.path.join(path,
                                '02 Data',
                                 'Original Data',
                                 'products.csv'))

In [4]:
# import df_ords
df_ords=pd.read_csv(os.path.join(path,
                                '02 Data',
                                'Prepared Data',
                                'orders_wrangled.csv'))

In [5]:
# Create dataframe

df_test=pd.DataFrame()

In [6]:
# Create a mixed type column

df_test['mix']=['a', 'b', 1, True]

In [7]:
# Show me

df_test

mix
0     a
1     b
2     1
3  True

In [8]:
# Function for checking a dataframe for mixed data types
# Returns the column name(s)

for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [9]:
# Convert 'mix' column to 'str'

df_test['mix']=df_test['mix'].astype('str')

In [10]:
# Find missing values
# Returns count of missing values in each column

df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [11]:
# Summon only the rows with NaN values in a new dataframe

df_nan=df_prods[df_prods['product_name'].isnull()==True]

In [12]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [13]:
# Count number of rows in df_prods and assign to variable

df_prodsRowCount1,columns=df_prods.shape
df_prodsRowCount1

49693

In [14]:
# Create new dataframe without 'NaN' values

df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [15]:
# Count number of rows in df_prods_clean and assign to variable

df_prodsRowCount2,columns=df_prods_clean.shape

In [16]:
# Find difference in row count between OG and cleaned variables

print(df_prodsRowCount1-df_prodsRowCount2)

16


In [17]:
# Great success!

In [18]:
# Create new subset containing only duplicates

df_dups=df_prods_clean[df_prods_clean.duplicated()]

In [19]:
# How many duplicates?

df_dups.shape

(5, 5)

In [20]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [21]:
# count rows

dpcs1,cols=df_prods_clean.shape

In [22]:
# Create new dataframe without duplicates that I indentified earlier

df_prods_clean_no_dups=df_prods_clean.drop_duplicates()

In [23]:
# count rows

dpcs2,cols=df_prods_clean_no_dups.shape

In [24]:
# Confirm difference

print(dpcs1-dpcs2)

5


In [25]:
# Export new dataframe

df_prods_clean_no_dups.to_csv(os.path.join(path,
                                          '02 Data',
                                          'Prepared Data',
                                          'products_checked.csv'))

# 1) Task 4.5 - Data Consistency Checks

## 2) Run the df.describe() function on your df_prods dataframe. Does anything look fishy?

In [26]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

### - 'prices' column has a max value of 99999, not good!

## 3) Check for mixed-type data in df_ords

In [27]:
# Function for checking a dataframe for mixed data types
# Returns the column name(s)

for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

### - didn't print any columns names, so it must be clean

## 4) Run a check for missing values in your df_ords dataframe.

In [28]:
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

### - In the 'days_since_prior_order' column, there are over 200,000 missing values. This could just mean that users ordered twice within 24 hours

In [29]:
df_ords_nan=df_ords[df_ords['days_since_prior_order'].isnull()==True]

In [30]:
df_ords_nan.head()

Unnamed: 0  order_id  user_id eval_set  order_number  orders_day_of_week  \
0            0   2539329        1    prior             1                   2   
11          11   2168274        2    prior             1                   2   
26          26   1374495        3    prior             1                   1   
39          39   3343014        4    prior             1                   6   
45          45   2717275        5    prior             1                   3   

    order_hour_of_day  days_since_prior_order  
0                   8                     NaN  
11                 11                     NaN  
26                 14                     NaN  
39                 11                     NaN  
45                 12                     NaN

In [31]:
# orders with order_id '2539329'

oiCheck1=df_ords[df_ords['order_id']==2539329]
oiCheck1

Unnamed: 0  order_id  user_id eval_set  order_number  orders_day_of_week  \
0           0   2539329        1    prior             1                   2   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN

In [32]:
# users with user_id '1'

oiCheck2=df_ords[df_ords['user_id']==1]
oiCheck2

Unnamed: 0  order_id  user_id eval_set  order_number  orders_day_of_week  \
0            0   2539329        1    prior             1                   2   
1            1   2398795        1    prior             2                   3   
2            2    473747        1    prior             3                   3   
3            3   2254736        1    prior             4                   4   
4            4    431534        1    prior             5                   4   
5            5   3367565        1    prior             6                   2   
6            6    550135        1    prior             7                   1   
7            7   3108588        1    prior             8                   1   
8            8   2295261        1    prior             9                   1   
9            9   2550362        1    prior            10                   4   
10          10   1187899        1    train            11                   4   

    order_hour_of_day  days_since_prior_order  
0                   8                     NaN  
1                   7                    15.0  
2                  12                    21.0  
3                   7                    29.0  
4                  15                    28.0  
5                   7                    19.0  
6                   9                    20.0  
7                  14                    14.0  
8                  16                     0.0  
9                   8                    30.0  
10                  8                    14.0

### - Users with NaN in 'days_since_prior_order' are new users?
#### - NOTE: there are 206209 NaN's in variable. Is this the same number of user_id's?

In [33]:
### Find number of unique user_id's

df_ords['user_id'].describe()

count    3.421083e+06
mean     1.029782e+05
std      5.953372e+04
min      1.000000e+00
25%      5.139400e+04
50%      1.026890e+05
75%      1.543850e+05
max      2.062090e+05
Name: user_id, dtype: float64

### Conclusion: Number of rows with NaN values in 'days_since_prior_order' is equal to the number of unique user_id's. I would infer that rows with this characteristic mark a customer's first order. 

## 5) Address the missing values using an appropriate method.

In [34]:
# df_ords_nan is useful for analyzing patterns for new customers
# Removed 'days_since_prior_order' column
# Subsetted as 'df_ords_new_cust'

df_ords_new_cust=df_ords_nan.drop(columns=
                                  ['days_since_prior_order'])
df_ords_new_cust.tail(5)

Unnamed: 0  order_id  user_id eval_set  order_number  \
3420930     3420930    969311   206205    prior             1   
3420934     3420934   3189322   206206    prior             1   
3421002     3421002   2166133   206207    prior             1   
3421019     3421019   2227043   206208    prior             1   
3421069     3421069   3154581   206209    prior             1   

         orders_day_of_week  order_hour_of_day  
3420930                   4                 12  
3420934                   3                 18  
3421002                   6                 19  
3421019                   1                 15  
3421069                   3                 11

In [35]:
# Get 'before' row count

df_ords1, cols = df_ords.shape

In [36]:
# I can remove these values, they represent > 1% of df_ords

df_ords_clean=df_ords[df_ords['days_since_prior_order'].isnull()==False]
df_ords_clean.head()

Unnamed: 0  order_id  user_id eval_set  order_number  orders_day_of_week  \
1           1   2398795        1    prior             2                   3   
2           2    473747        1    prior             3                   3   
3           3   2254736        1    prior             4                   4   
4           4    431534        1    prior             5                   4   
5           5   3367565        1    prior             6                   2   

   order_hour_of_day  days_since_prior_order  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0  
5                  7                    19.0

In [37]:
# Get 'after' row count

df_ords2, cols = df_ords_clean.shape

In [38]:
# Get difference between before and after...

df_ords1-df_ords2

206209

### - I decided to remove the null values because they represent > 1% of df_ords. I placed the nulls in another dataset, thinking that they may be useful for analysis of new customer orders.

# 6) Run a check for duplicate values in your df_ords data.

In [39]:
# Find full duplicates in 'df_ords'

df_ords_dups=df_ords_clean[df_ords_clean.duplicated()]
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

### - There appears to be no duplicates

# 7) Exporting

In [40]:
# Export 'df_ords_clean'

df_ords_clean.to_csv(os.path.join(path,
                                 '02 Data',
                                 'Prepared Data',
                                 'orders_checked.csv'))

In [41]:
# Export 'df_ords_new_cust'

df_ords_new_cust.to_csv(os.path.join(path,
                                 '02 Data',
                                 'Prepared Data',
                                 'new_customer_orders.csv'))

In [42]:
# Exported 'df_prods_clean_no_dups' earlier in notebook as 'products_checked.csv'